src/concurrent-queue.js

```diff
@@ -0,0 +1,172 @@
+ *   when additional items are added.
+ */
+export default class ConcurrentQueue extends UploadOptionsBase {
+
+    /**
+     * Constructs a new instance of the queue.
+     * @param {object} options Used by the super class.
+     * @param {DirectBinaryUploadOptions} uploadOptions Used to determine
+     *  the maximum number of items to process concurrently. Retrieved using
+     *  the options' getMaxConcurrent() method.
+     */
+    constructor(options, uploadOptions) {
+        super(options, uploadOptions);
+        this.queue = Async.queue(async (task, callback) => {
+            const {
+                data,
+                worker,
+                batch,
+            } = task;
+
+            let err = undefined;
+            try {
+                await worker(data);
+            } catch (e) {
+                err = e;
+            }
+
+            updateBatch.call(this, batch);
+
+            callback(err);
+        }, uploadOptions.getMaxConcurrent());
+
+        // send 'error' event if encountering an error while
+        // processing an item.
+        this.queue.error((error, task) => {
+            const { data = {} } = task;
+            this.emit('error', {
+                item: data,
+                error,
+            });
+        });
+
+        // will be sent whenever all items added to the queue
+        // have processed. may be sent multiple times throughout
+        // the life of the queue.
+        this.queue.drain(() => {
+            this.emit('emptied', true);
+        });
+
+        this.batchCounts = {};
+    }
+
+    /**
+     * Adds a new item to the queue for processing.
+     * @param {*} item The item to process.
+     * @param {function} callback Will be invoked when the item is ready to
+     *  be processed. The function is expected to return a promise. It will be
+     *  invoked with a single argument: the item parameter provided in this method.
+     * @returns {Promise} Will be resolved when the item has been processed.
+     */
+    push(item, callback) {
+        return new Promise((res) => {
+            this.queue.push({
+                data: item,
+                batch: createBatch.call(this, 1, res),
+                worker: callback,
+            });
+        });
+    }
+
+    /**
+     * Adds an array of items to the queue for processing. Each item will be
+     * passed to the same processing callback.
+     * @param {Array} items Array of items to process.
+     * @param {function} callback Will be invoked when one of item is ready to
+     *  be processed. The function is expected to return a promise. It will be
+     *  invoked with a single argument: the item to be processed.
+     * @returns {Promise} Will be resolved when all provided items have been processed.
+     */
+    pushAll(items, callback) {
+        return new Promise((res) => {
+            const batch = createBatch.call(this, items.length, res);
+            items.forEach((item) => {
```

Comment:
Ugh, javascript. I sure do miss strongly typed languages :)

I totally see where you're coming from, but in a case like this it seems like we have a couple options after verifying:

1. Ignore the value and silently do nothing. I don't think this is a good option because it would "hide" that the caller is doing something wrong. Better to tell them upfront that the parameter is wrong.
2. Throw an exception indicating that the parameter needs to be iterable. This seems better because we're being more verbose about what went wrong.

However, as it's written there would be an exception along the lines of `forEach is not a function`. IMO, that exception combined with the method documentation indicating that the parameter needs to be an `Array` gives the caller enough information to know what the problem is. And this way we're not bloating the method with a bunch of validation code; because if we're going to verify that `items` is iterable, then we should probably also verify that `callback` is a function.

WDYT? Are there reasons for doing the validation that I'm missing?